x/challenge/keeper/msg_server_submit.go

```diff
@@ -18,52 +18,72 @@ func (k msgServer) Submit(goCtx context.Context, msg *types.MsgSubmit) (*types.M
 	challenger := sdk.MustAccAddressFromHex(msg.Challenger)
 
 	// check sp status
-	sp, found := k.SpKeeper.GetStorageProvider(ctx, spOperator)
+	bucketInfo, found := k.StorageKeeper.GetBucketInfo(ctx, msg.BucketName)
+	if !found {
+		return nil, types.ErrUnknownBucketObject
+	}
+	sp, found := k.SpKeeper.GetStorageProvider(ctx, bucketInfo.PrimarySpId)
 	if !found {
 		return nil, types.ErrUnknownSp
 	}
-	if sp.Status != sptypes.STATUS_IN_SERVICE {
+	if sp.Status != sptypes.STATUS_IN_SERVICE && sp.Status != sptypes.STATUS_GRACEFUL_EXITING {
 		return nil, types.ErrInvalidSpStatus
 	}
 
 	// check object & read needed data
 	objectInfo, found := k.StorageKeeper.GetObjectInfo(ctx, msg.BucketName, msg.ObjectName)
 	if !found {
-		return nil, types.ErrUnknownObject
+		return nil, types.ErrUnknownBucketObject
 	}
 	if objectInfo.ObjectStatus != storagetypes.OBJECT_STATUS_SEALED {
 		return nil, types.ErrInvalidObjectStatus
 	}
 
-	// check whether the sp stores the object info
+	// check whether the sp stores the object info, generate redundancy index
 	stored := false
-	for _, sp := range objectInfo.GetSecondarySpAddresses() {
-		if spOperator.Equals(sdk.MustAccAddressFromHex(sp)) {
-			stored = true
-			break
-		}
+	redundancyIndex := types.RedundancyIndexPrimary
+
+	// check primary sp
+	tmpSp, found := k.SpKeeper.GetStorageProvider(ctx, bucketInfo.PrimarySpId)
+	if !found {
+		return nil, types.ErrFailToSubmit
+	}
+	if spOperator.Equals(sdk.MustAccAddressFromHex(tmpSp.OperatorAddress)) {
+		stored = true
 	}
+
 	if !stored {
-		bucket, _ := k.StorageKeeper.GetBucketInfo(ctx, msg.BucketName)
-		if !spOperator.Equals(sdk.MustAccAddressFromHex(bucket.GetPrimarySpAddress())) {
-			return nil, types.ErrNotStoredOnSp
+		lvg, found := k.VirtualGroupKeeper.GetLVG(ctx, bucketInfo.Id, objectInfo.LocalVirtualGroupId)
+		if !found {
+			return nil, types.ErrFailToSubmit
+		}
+		gvg, found := k.VirtualGroupKeeper.GetGVG(ctx, lvg.GlobalVirtualGroupId)
+		if !found {
+			return nil, types.ErrFailToSubmit
```

Comment:
`ErrFailToSubmit` is too common error.

Reply:
refactor to use more specific error code